### PR TITLE
add metadata project

### PR DESCRIPTION
**Pull request by vwxyzh**
use metadata schema to validate metadata, or generate strong type.

### CODE REVIEW

File: src/Microsoft.DocAsCode.Metadata/DisplayNameAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿namespace Microsoft.DocAsCode.Metadata
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Metadata/DisplayNameAttribute.cs]: Copyright header.

File: All.sln [hunk 2]
```diff
@@ -383,6 +383,12 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.DocAsCode.Common"
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.DocAsCode.MarkdownRewriters", "src\Microsoft.DocAsCode.MarkdownRewriters\Microsoft.DocAsCode.MarkdownRewriters.xproj", "{5FF9B54C-A853-4FA8-A771-40837D062F05}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.DocAsCode.Metadata", "src\Microsoft.DocAsCode.Metadata\Microsoft.DocAsCode.Metadata.csproj", "{AC1D8334-F12A-846A-3432-B6F9FA45C533}"
```
Comment (vicancy) [inline All.sln]: is the naming too general?

File: src/Microsoft.DocAsCode.Metadata/SchemaValidators/UnknownNamingValidator.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+        public ValidationResult Validate(string name, JToken value)
+        {
+            if (Regex.IsMatch(name))
+            {
+                return ValidationResult.Success;
+            }
+            else
+            {
+                return ValidationResult.Fail(ValidationErrorCodes.UnknownMetadata.BadNaming, "Bad naming for unknown metadata {name}.", name);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Metadata/SchemaValidators/UnknownNamingValidator.cs]: Details about the criteria of supported naming?

File: src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.DocAsCode.Metadata.SchemaValidators
+{
+    public static class ValidatorFactory
+    {
+        public static IWellknownMetadataValidator Then(this IWellknownMetadataValidator validator, IWellknownMetadataValidator next)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs]: null check?

File: src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                return new ChainWellknownValidator
+                {
+                    Validators =
+                    {
+                        validator,
+                        next,
+                    }
+                };
+            }
+            chain.Validators.Add(next);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs]: thread safety?

File: src/Microsoft.DocAsCode.Metadata/ValidationResults.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+    public class ValidationResults
+    {
+        public ValidationResults(IEnumerable<ValidationResult> results)
+        {
+            Items.AddRange(from r in results where !r.IsSuccess select r);
+        }
+
+        public bool IsSuccess => Items.Count == 0;
+        public List<ValidationResult> Items { get; } = new List<ValidationResult>();
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Metadata/ValidationResults.cs]: => is enough?

File: All.sln (hunk 2 above)
Comment (vwxyzh) [inline All.sln]: rename to MetadataSchema

File: src/Microsoft.DocAsCode.Metadata/ValidationResults.cs (hunk 6 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata/ValidationResults.cs]: '=>' will eval every time access the property, '=' will only eval in constructor.

File: src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs (hunk 5 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs]: instance member is not thread safety here.

File: src/Microsoft.DocAsCode.Metadata/SchemaValidators/UnknownNamingValidator.cs (hunk 3 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata/SchemaValidators/UnknownNamingValidator.cs]: start with a-z, following with a-z, 0-9, and _

File: src/Microsoft.DocAsCode.Metadata/DisplayNameAttribute.cs (hunk 1 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata/DisplayNameAttribute.cs]: fixed.

File: src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs (hunk 4 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata/SchemaValidators/ValidatorFactory.cs]: added.

**Comment by vicancy** [comment]
LGTM. :+1:

**Pull request closed without merging**